From 4813922ae84c0485f8a2e909e40bc84cf279edb1 Mon Sep 17 00:00:00 2001
From: Liu Yafei <yfliu@mobvoi.com>
Date: Sat, 16 Jul 2016 13:36:04 +0800
Subject: [PATCH] When need DRAW_WAKE_LOCK, checks device wakefullness instead
 of checking display status.

There's a latency between the doze state in PowerManagerSerivce and the DisplayManagerService,
this will occur a different behavior in some extreme case when system is dozing but the display
state is ON. The normal appearance when this happends is that while system is dozing and display
is ON, a view will not hold a draw wake lock and do the drawing, but when the drawing command is
sent to hwcomposer, the display state has just turn to DOZE_SUSPEND due to no wake lock is holding
so that the hwcomposer will not perform the drawing action.

This situation was mentioned by jeffbrown@google.com in the first place when pokeDrawLock() was
written.

Change-Id: Ica10fe8b021c0cbfa459eee1ef82bf1938e0bf1c
---
 core/java/android/service/wallpaper/WallpaperService.java | 8 ++++++--
 core/java/android/view/ViewRootImpl.java                  | 7 ++++---
 2 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/core/java/android/service/wallpaper/WallpaperService.java b/core/java/android/service/wallpaper/WallpaperService.java
index 1863335..5eb216f 100644
--- a/core/java/android/service/wallpaper/WallpaperService.java
+++ b/core/java/android/service/wallpaper/WallpaperService.java
@@ -18,6 +18,7 @@ package android.service.wallpaper;
 
 import android.content.res.TypedArray;
 import android.os.Build;
+import android.os.PowerManager;
 import android.os.SystemProperties;
 import android.graphics.Canvas;
 import android.util.DisplayMetrics;
@@ -196,6 +197,8 @@ public abstract class WallpaperService extends Service {
         Display mDisplay;
         private int mDisplayState;
 
+        PowerManager mPowerManager;
+
         final BaseSurfaceHolder mSurfaceHolder = new BaseSurfaceHolder() {
             {
                 mRequestedFormat = PixelFormat.RGBX_8888;
@@ -241,8 +244,7 @@ public abstract class WallpaperService extends Service {
 
             @Override
             public Canvas lockCanvas() {
-                if (mDisplayState == Display.STATE_DOZE
-                        || mDisplayState == Display.STATE_DOZE_SUSPEND) {
+                if (mPowerManager.isInteractive()) {
                     try {
                         mSession.pokeDrawLock(mWindow);
                     } catch (RemoteException e) {
@@ -887,6 +889,8 @@ public abstract class WallpaperService extends Service {
             mDisplay = mDisplayManager.getDisplay(Display.DEFAULT_DISPLAY);
             mDisplayState = mDisplay.getState();
 
+            mPowerManager = (PowerManager) getSystemService(Context.POWER_SERVICE);
+
             if (DEBUG) Log.v(TAG, "onCreate(): " + this);
             onCreate(mSurfaceHolder);
             
diff --git a/core/java/android/view/ViewRootImpl.java b/core/java/android/view/ViewRootImpl.java
index efcf6ed..53caa62 100644
--- a/core/java/android/view/ViewRootImpl.java
+++ b/core/java/android/view/ViewRootImpl.java
@@ -47,6 +47,7 @@ import android.os.Handler;
 import android.os.Looper;
 import android.os.Message;
 import android.os.ParcelFileDescriptor;
+import android.os.PowerManager;
 import android.os.Process;
 import android.os.RemoteException;
 import android.os.SystemClock;
@@ -149,6 +150,7 @@ public final class ViewRootImpl implements ViewParent,
     final IWindowSession mWindowSession;
     final Display mDisplay;
     final DisplayManager mDisplayManager;
+    final PowerManager mPowerManager;
     final String mBasePackageName;
 
     final int[] mTmpLocation = new int[2];
@@ -417,6 +419,7 @@ public final class ViewRootImpl implements ViewParent,
         mFallbackEventHandler = PolicyManager.makeNewFallbackEventHandler(context);
         mChoreographer = Choreographer.getInstance();
         mDisplayManager = (DisplayManager)context.getSystemService(Context.DISPLAY_SERVICE);
+        mPowerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
         loadSystemProperties();
         mWindowIsRound = context.getResources().getBoolean(
                 com.android.internal.R.bool.config_windowIsRound);
@@ -980,10 +983,8 @@ public final class ViewRootImpl implements ViewParent,
     };
 
     void pokeDrawLockIfNeeded() {
-        final int displayState = mAttachInfo.mDisplayState;
         if (mView != null && mAdded && mTraversalScheduled
-                && (displayState == Display.STATE_DOZE
-                        || displayState == Display.STATE_DOZE_SUSPEND)) {
+                && (mPowerManager.isInteractive())) {
             try {
                 mWindowSession.pokeDrawLock(mWindow);
             } catch (RemoteException ex) {
-- 
1.9.1

